### PR TITLE
CI: Intel ICX/ICPX

### PR DESCRIPTION
**Pull request by ax3l**
Add coverage for Intels icx/icpx compilers.
Those are the next-gen intel compilers, while icc/icpc are considered "classic" (read as: legacy).

### CODE REVIEW

File: .github/workflows/ubuntu.yml [hunk 1]
```diff
@@ -78,6 +78,54 @@ jobs:
...
+    - name: install dependencies
+      run: |
+        export DEBIAN_FRONTEND=noninteractive
+        sudo apt-get -qqq update
+        sudo apt-get install -y wget build-essential pkg-config cmake ca-certificates gnupg
+        sudo wget https://apt.repos.intel.com/intel-gpg-keys/GPG-PUB-KEY-INTEL-SW-PRODUCTS-2023.PUB
+        sudo apt-key add GPG-PUB-KEY-INTEL-SW-PRODUCTS-2023.PUB
+        echo "deb https://apt.repos.intel.com/oneapi all main" | sudo tee /etc/apt/sources.list.d/oneAPI.list
+        sudo apt-get update
+        sudo apt-get install -y intel-oneapi-compiler-dpcpp-cpp
```
Comment (ax3l) [inline .github/workflows/ubuntu.yml]: Missing declared package dependency?
```suggestion
        sudo apt-get install -y intel-oneapi-compiler-dpcpp-cpp intel-oneapi-compiler-shared-runtime-2021.1.1
```

File: .github/workflows/ubuntu.yml [hunk 2]
```diff
@@ -78,6 +78,54 @@ jobs:
...
+    - name: install dependencies
+      run: |
+        export DEBIAN_FRONTEND=noninteractive
+        sudo apt-get -qqq update
+        sudo apt-get install -y wget build-essential pkg-config cmake ca-certificates gnupg
+        sudo wget https://apt.repos.intel.com/intel-gpg-keys/GPG-PUB-KEY-INTEL-SW-PRODUCTS-2023.PUB
+        sudo apt-key add GPG-PUB-KEY-INTEL-SW-PRODUCTS-2023.PUB
+        echo "deb https://apt.repos.intel.com/oneapi all main" | sudo tee /etc/apt/sources.list.d/oneAPI.list
+        sudo apt-get update
+        sudo apt-get install -y intel-oneapi-compiler-dpcpp-cpp intel-oneapi-compiler-shared-runtime-2021.1.1
```
Comment (ax3l) [inline .github/workflows/ubuntu.yml]: Trying more explicit:
```suggestion
        sudo apt-get install -y intel-oneapi-dpcpp-cpp-2021.1.2 intel-oneapi-compiler-shared-runtime-2021.1.2
```

File: .github/workflows/ubuntu.yml [hunk 3]
```diff
@@ -78,6 +78,54 @@ jobs:
...
+        set +e
+        source /opt/intel/oneapi/setvars.sh
+        set -e
+        sudo curl -L -o /usr/local/bin/cmake-easyinstall https://git.io/JvLxY
+        sudo chmod a+x /usr/local/bin/cmake-easyinstall
+        export CEI_SUDO="sudo"
+        CXX=$(which icpx) CC=$(which icx) cmake-easyinstall --prefix=/usr/local git+https://github.com/openPMD/openPMD-api.git -DopenPMD_USE_PYTHON=OFF -DBUILD_TESTING=OFF -DBUILD_EXAMPLES=OFF \
+          -DCMAKE_CXX_COMPILER_ID="Clang"                \
+          -DCMAKE_CXX_COMPILER_VERSION=12.0              \
+          -DCMAKE_CXX_STANDARD_COMPUTED_DEFAULT="17"     \
```
Comment (ax3l) [inline .github/workflows/ubuntu.yml]: Remove unnecessary line continuation
```suggestion
          -DCMAKE_CXX_STANDARD_COMPUTED_DEFAULT="17"
```

File: .github/workflows/ubuntu.yml [hunk 4]
```diff
@@ -78,6 +78,54 @@ jobs:
...
+        set +e
+        source /opt/intel/oneapi/setvars.sh
+        set -e
+        export CXX=$(which icpx)
+        export CC=$(which icx)
+
+        cmake -S . -B build_dp -DCMAKE_VERBOSE_MAKEFILE=ON -DWarpX_MPI=OFF -DWarpX_OPENPMD=ON -DWarpX_openpmd_internal=OFF \
+          -DCMAKE_CXX_COMPILER_ID="Clang"                \
+          -DCMAKE_CXX_COMPILER_VERSION=12.0              \
+          -DCMAKE_CXX_STANDARD_COMPUTED_DEFAULT="17"     \
```
Comment (ax3l) [inline .github/workflows/ubuntu.yml]:
```suggestion
          -DCMAKE_CXX_STANDARD_COMPUTED_DEFAULT="17"
```

File: .github/workflows/ubuntu.yml [hunk 5]
```diff
@@ -78,6 +78,54 @@ jobs:
...
+        cmake --build build_dp -j 2
+
+        cmake -S . -B build_sp -DCMAKE_VERBOSE_MAKEFILE=ON -DWarpX_MPI=OFF -DWarpX_OPENPMD=ON -DWarpX_openpmd_internal=OFF -DWarpX_PRECISION=SINGLE \
+          -DCMAKE_CXX_COMPILER_ID="Clang"                \
+          -DCMAKE_CXX_COMPILER_VERSION=12.0              \
+          -DCMAKE_CXX_STANDARD_COMPUTED_DEFAULT="17"     \
```
Comment (ax3l) [inline .github/workflows/ubuntu.yml]:
```suggestion
          -DCMAKE_CXX_STANDARD_COMPUTED_DEFAULT="17"
```

File: .github/workflows/ubuntu.yml [hunk 6]
```diff
@@ -78,6 +78,54 @@ jobs:
...
+    - name: install dependencies
+      run: |
+        export DEBIAN_FRONTEND=noninteractive
+        sudo apt-get -qqq update
+        sudo apt-get install -y wget build-essential pkg-config cmake ca-certificates gnupg
+        sudo wget https://apt.repos.intel.com/intel-gpg-keys/GPG-PUB-KEY-INTEL-SW-PRODUCTS-2023.PUB
+        sudo apt-key add GPG-PUB-KEY-INTEL-SW-PRODUCTS-2023.PUB
+        echo "deb https://apt.repos.intel.com/oneapi all main" | sudo tee /etc/apt/sources.list.d/oneAPI.list
+        sudo apt-get update
+        sudo apt-get install -y intel-oneapi-dpcpp-cpp-2021.1.2 intel-oneapi-compiler-shared-runtime-2021.1.2
```
Comment (ax3l) [inline .github/workflows/ubuntu.yml]: Throwing in more dependencies as in dpcpp:
```suggestion
        sudo apt-get install -y intel-oneapi-dpcpp-cpp-2021.1.2 intel-oneapi-compiler-shared-runtime-2021.1.2 build-essential cmake intel-oneapi-dpcpp-cpp-compiler intel-oneapi-mkl-devel
```

File: .github/workflows/ubuntu.yml [hunk 7]
```diff
@@ -78,6 +78,42 @@ jobs:
...
+      shell: bash
+      run: |
+        set +e
+        source /opt/intel/oneapi/setvars.sh
+        set -e
+        export CXX=$(which icpx)
+        export CC=$(which icx)
+
+        cmake -S . -B build_sp         \
+          -DCMAKE_CXX_COMPILER_ID="Clang"                \
```
Comment (ax3l) [inline .github/workflows/ubuntu.yml:103]: More work-arounds for broken Intel patch that landed in oneAPI:
```suggestion
          -DCMAKE_C_COMPILER_ID="Clang"                  \
          -DCMAKE_C_COMPILER_VERSION=12.0                \ 
          -DCMAKE_CXX_COMPILER_ID="Clang"                \
```

File: .github/workflows/ubuntu.yml [hunk 8]
```diff
@@ -78,6 +78,44 @@ jobs:
...
+      run: |
+        set +e
+        source /opt/intel/oneapi/setvars.sh
+        set -e
+        export CXX=$(which icpx)
+        export CC=$(which icx)
+
+        cmake -S . -B build_sp         \
+          -DCMAKE_C_COMPILER_ID="Clang"                  \
+          -DCMAKE_C_COMPILER_VERSION=12.0                \
```
Comment (ax3l) [inline .github/workflows/ubuntu.yml]: EOL whitespace
```suggestion
          -DCMAKE_C_COMPILER_VERSION=12.0                \
```

File: .github/workflows/ubuntu.yml [hunk 9]
```diff
@@ -78,6 +78,44 @@ jobs:
...
+          -DCMAKE_CXX_COMPILER_ID="Clang"                \
```
Comment (ax3l) [inline .github/workflows/ubuntu.yml:103]:
```suggestion
          -DCMAKE_CXX_COMPILER_ID="Clang"                \
          -DCMAKE_CXX_STANDARD_COMPUTED_DEFAULT="11"     \
```

File: .github/workflows/ubuntu.yml [hunk 10]
```diff
@@ -78,6 +78,45 @@ jobs:
...
+          -DCMAKE_CXX_STANDARD_COMPUTED_DEFAULT="11"     \
```
Comment (ax3l) [inline .github/workflows/ubuntu.yml]:
```suggestion
          -DCMAKE_C_STANDARD_COMPUTED_DEFAULT="11"     \
          -DCMAKE_CXX_COMPILER_ID="Clang"                \
```

File: .github/workflows/ubuntu.yml [hunk 11]
```diff
@@ -78,6 +78,45 @@ jobs:
...
+          -DCMAKE_CXX_COMPILER_VERSION=12.0              \
+          -DCMAKE_CXX_STANDARD_COMPUTED_DEFAULT="17"     \
```
Comment (ax3l) [inline .github/workflows/ubuntu.yml]:
```suggestion
          -DCMAKE_CXX_STANDARD_COMPUTED_DEFAULT="14"     \
```

File: .github/workflows/ubuntu.yml [hunk 12]
```diff
@@ -78,6 +78,45 @@ jobs:
...
+
+        cmake -S . -B build_sp         \
+          -DCMAKE_C_COMPILER_ID="Clang"                  \
+          -DCMAKE_C_COMPILER_VERSION=12.0                \
+          -DCMAKE_C_STANDARD_COMPUTED_DEFAULT="11"     \
+          -DCMAKE_CXX_COMPILER_ID="Clang"                \
+          -DCMAKE_CXX_COMPILER_VERSION=12.0              \
+          -DCMAKE_CXX_STANDARD_COMPUTED_DEFAULT="14"     \
+          -DCMAKE_VERBOSE_MAKEFILE=ON  \
+          -DWarpX_COMPUTE=SYCL         \
```
Comment (ax3l) [inline .github/workflows/ubuntu.yml]:
```suggestion
```

File: .github/workflows/ubuntu.yml [hunk 13]
```diff
@@ -78,6 +78,45 @@ jobs:
...
+        set +e
+        source /opt/intel/oneapi/setvars.sh
+        set -e
+        export CXX=$(which icpx)
+        export CC=$(which icx)
+
+        cmake -S . -B build_sp         \
+          -DCMAKE_C_COMPILER_ID="Clang"                  \
+          -DCMAKE_C_COMPILER_VERSION=12.0                \
+          -DCMAKE_C_STANDARD_COMPUTED_DEFAULT="11"     \
```
Comment (ax3l) [inline .github/workflows/ubuntu.yml]:
```suggestion
          -DCMAKE_C_STANDARD_COMPUTED_DEFAULT="11"       \
```

**Review by RemiLehe — APPROVED** [review]

**Pull request merged**